sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/StorageBlobServiceImpl.java

```diff
@@ -562,6 +564,127 @@ ServiceCall stageBlockWithRestResponse(String containerName,
         return new ServiceCall(callAndOptionalResult.getCall());
     }
 
+    /**
+     * Undelete a blob that was previously soft-deleted.
+     * <p>
+     * Undelete restores the content and metadata of a soft-deleted blob and/or any associated soft-deleted snapshots.
+     *
+     * @param containerName The container name.
+     * @param blobName      The blob name.
+     */
+    Void undelete(String containerName,
+                  String blobName) {
+        return undeleteWithResponse(containerName,
+            blobName,
+            null,
+            null,
+            null,
+            null,
+            null).getValue();
+    }
+
+    /**
+     * Undelete a blob that was previously soft-deleted.
+     * <p>
+     * Undelete restores the content and metadata of a soft-deleted blob and/or any associated soft-deleted snapshots.
+     *
+     * @param containerName The container name.
+     * @param blobName      The blob name.
+     * @param callback      Callback that receives the response.
+     * @return A handle to the service call.
+     */
+    ServiceCall undelete(String containerName,
+                         String blobName,
+                         Callback<Void> callback) {
+        return undeleteWithResponse(containerName,
+            blobName,
+            null,
+            null,
+            null,
+            null,
+            "undelete",
+            new Callback<BlobUndeleteResponse>() {
+                @Override
+                public void onResponse(BlobUndeleteResponse response) {
+                    callback.onResponse(response.getValue());
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    callback.onFailure(t);
+                }
+            });
+    }
+
+    /**
+     * Undelete a blob that was previously soft-deleted.
+     * <p>
+     * Undelete restores the content and metadata of a soft-deleted blob and/or any associated soft-deleted snapshots.
+     *
+     * @param containerName The container name.
+     * @param blobName      The blob name.
+     * @param url           The URL to the deleted blob.
+     * @param timeout       The timeout parameter is expressed in seconds. For more information, see
+     *                      &lt;a href="https://docs.microsoft.com/en-us/rest/api/storageservices/fileservices/setting-timeouts-for-blob-service-operations"&gt;Setting Timeouts for Blob Service Operations.&lt;/a&gt;.
+     * @param requestId     Provides a client-generated, opaque value with a 1 KB character limit that is
+     *                      recorded in the analytics logs when storage analytics logging is enabled.
+     * @param comp          The name of the desired operation to perform. In this case, it should be 'undelete'.
+     * @return A response object containing the details of the delete operation.
+     */
+    BlobUndeleteResponse undeleteWithResponse(String containerName,
```

Comment:
This should not take "comp" as a parameter. For undelete it must always be the string "undelete".

Reply:
I suspect the right place to hard-code the value for "comp" is actually in `undeleteWithRestResponseIntern`